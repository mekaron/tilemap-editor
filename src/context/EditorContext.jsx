import React, { createContext, useReducer, useContext } from 'react';

export const TOOLS = {
  BRUSH: 0,
  ERASE: 1,
  PAN: 2,
  PICK: 3,
  RAND: 4,
  FILL: 5,
};

export const initialState = {
  prevActiveTool: TOOLS.BRUSH,
  activeTool: TOOLS.BRUSH,
  activeMap: '',
  displaySymbols: false,
  showGrid: false,
  selection: [],
  currentLayer: 0,
  isMouseDown: false,
  maps: {},
  tileSets: {},
  activeLoader: 'default',
};

export const editorReducer = (state, action) => {
  switch (action.type) {
    case 'SET_TOOL':
      return {
        ...state,
        prevActiveTool: state.activeTool,
        activeTool: action.tool,
      };
    case 'ADD_MAP':
      return {
        ...state,
        maps: { ...state.maps, [action.map.name]: action.map },
        activeMap: action.map.name,
      };
    case 'UPDATE_LAYER': {
      const { mapName, layerIndex, layerData } = action;
      const map = state.maps[mapName];
      if (!map) return state;
      const layers = map.layers.map((l, i) =>
        i === layerIndex ? { ...l, ...layerData } : l
      );
      return {
        ...state,
        maps: { ...state.maps, [mapName]: { ...map, layers } },
      };
    }
    case 'SET_ACTIVE_MAP':
      return { ...state, activeMap: action.name };
    case 'TOGGLE_GRID':
      return { ...state, showGrid: !state.showGrid };
    default:
      return state;
  }
};

const EditorContext = createContext();

export const EditorProvider = ({ children }) => {
  const [state, dispatch] = useReducer(editorReducer, initialState);
  return (
    <EditorContext.Provider value={{ state, dispatch, TOOLS }}>
      {children}
    </EditorContext.Provider>
  );
};

export const useEditor = () => useContext(EditorContext);

export default EditorContext;
