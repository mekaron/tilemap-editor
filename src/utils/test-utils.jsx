import React from 'react';
import { render } from '@testing-library/react';
import EditorProvider from '../context/EditorProvider';

const AllTheProviders = ({ children }) => {
  return <EditorProvider>{children}</EditorProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
