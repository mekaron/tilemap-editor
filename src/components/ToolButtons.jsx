import React, { useContext } from 'react';
import EditorContext from '../context/EditorContext';

export default function ToolButtons() {
  const { editorState, setEditorState } = useContext(EditorContext);
  const { activeTool } = editorState;

  const handleToolChange = (e) => {
    setEditorState({ ...editorState, activeTool: Number(e.target.value) });
  };

  return (
    <div id="toolButtonsWrapper" className="tool_wrapper">
      <input id="tool0" type="radio" value="0" name="tool" checked={activeTool === 0} onChange={handleToolChange} className="hidden" />
      <label htmlFor="tool0" title="paint tiles" data-value="0" className="menu">
        <div id="flipBrushIndicator">🖌️</div>
        <div className="dropdown">
          <div className="item nohover">Brush tool options</div>
          <div className="item">
            <label htmlFor="toggleFlipX" className="">Flip tile on x</label>
            <input type="checkbox" id="toggleFlipX" style={{ display: 'none' }} />
            <label className="toggleFlipX"></label>
          </div>
        </div>
      </label>
      <input id="tool1" type="radio" value="1" name="tool" checked={activeTool === 1} onChange={handleToolChange} className="hidden" />
      <label htmlFor="tool1" title="erase tiles" data-value="1">🗑️</label>
      <input id="tool2" type="radio" value="2" name="tool" checked={activeTool === 2} onChange={handleToolChange} className="hidden" />
      <label htmlFor="tool2" title="pan" data-value="2">✋</label>
      <input id="tool3" type="radio" value="3" name="tool" checked={activeTool === 3} onChange={handleToolChange} className="hidden" />
      <label htmlFor="tool3" title="pick tile" data-value="3">🎨</label>
      <input id="tool4" type="radio" value="4" name="tool" checked={activeTool === 4} onChange={handleToolChange} className="hidden" />
      <label htmlFor="tool4" title="random from selected" data-value="4">🎲</label>
      <input id="tool5" type="radio" value="5" name="tool" checked={activeTool === 5} onChange={handleToolChange} className="hidden" />
      <label htmlFor="tool5" title="fill on layer" data-value="5">🌈</label>
    </div>
  );
}
