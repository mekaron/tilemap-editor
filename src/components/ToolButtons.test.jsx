import React from 'react';
import { render, screen, fireEvent } from '../utils/test-utils';
import ToolButtons from './ToolButtons';
import EditorContext from '../context/EditorContext';

const mockEditorState = {
  activeTool: 0,
};

const setEditorState = jest.fn();

const renderComponent = () =>
  render(
    <EditorContext.Provider value={{ editorState: mockEditorState, setEditorState }}>
      <ToolButtons />
    </EditorContext.Provider>
  );

test('ToolButtons renders correctly', () => {
  const { container } = renderComponent();
  expect(screen.getByTitle('paint tiles')).toBeTruthy();
  expect(container.querySelector('#toolButtonsWrapper')).toMatchSnapshot();
});

test('ToolButtons changes active tool on click', () => {
  renderComponent();
  const eraseButton = screen.getByTitle('erase tiles');
  fireEvent.click(eraseButton);
  expect(setEditorState).toHaveBeenCalledWith({
    ...mockEditorState,
    activeTool: 1,
  });
});
