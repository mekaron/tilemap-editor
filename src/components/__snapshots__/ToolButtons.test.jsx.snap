// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolButtons renders correctly 1`] = `
<div
  class="tool_wrapper"
  id="toolButtonsWrapper"
>
  <input
    checked=""
    class="hidden"
    id="tool0"
    name="tool"
    type="radio"
    value="0"
  />
  <label
    class="menu"
    data-value="0"
    for="tool0"
    title="paint tiles"
  >
    <div
      id="flipBrushIndicator"
    >
      🖌️
    </div>
    <div
      class="dropdown"
    >
      <div
        class="item nohover"
      >
        Brush tool options
      </div>
      <div
        class="item"
      >
        <label
          class=""
          for="toggleFlipX"
        >
          Flip tile on x
        </label>
        <input
          id="toggleFlipX"
          style="display: none;"
          type="checkbox"
        />
        <label
          class="toggleFlipX"
        />
      </div>
    </div>
  </label>
  <input
    class="hidden"
    id="tool1"
    name="tool"
    type="radio"
    value="1"
  />
  <label
    data-value="1"
    for="tool1"
    title="erase tiles"
  >
    🗑️
  </label>
  <input
    class="hidden"
    id="tool2"
    name="tool"
    type="radio"
    value="2"
  />
  <label
    data-value="2"
    for="tool2"
    title="pan"
  >
    ✋
  </label>
  <input
    class="hidden"
    id="tool3"
    name="tool"
    type="radio"
    value="3"
  />
  <label
    data-value="3"
    for="tool3"
    title="pick tile"
  >
    🎨
  </label>
  <input
    class="hidden"
    id="tool4"
    name="tool"
    type="radio"
    value="4"
  />
  <label
    data-value="4"
    for="tool4"
    title="random from selected"
  >
    🎲
  </label>
  <input
    class="hidden"
    id="tool5"
    name="tool"
    type="radio"
    value="5"
  />
  <label
    data-value="5"
    for="tool5"
    title="fill on layer"
  >
    🌈
  </label>
</div>
`;
